File: lib/services/object_engine.dart
import 'dart:ui';

import 'package:flutter/foundation.dart';
import 'package:ultralytics_yolo/ultralytics_yolo.dart';

import '../models/detection.dart';

/// 物体识别引擎：封装官方 Ultralytics YOLO 插件（YOLO11）。
///
/// **为什么不卡顿**：YOLO 推理在原生线程（平台通道）执行，不占用 Dart 主
/// isolate；调用方（[AppController]）以「后台节流 + JPEG 编码放 Isolate」的方式
/// 驱动它，绝不阻塞人脸/注视的高频帧循环。
///
/// **识别能力**：YOLO11 检测 + COCO 80 类，含 cup / bottle / cell phone / book /
/// laptop / remote / banana 等常见手持物体，定位与准确率均优于「整图分类器套裁剪
/// 框」的旧方案。标签经 [_cocoZh] 映射为中文，未命中返回英文原标签。
///
/// **坐标空间对齐**：[predictJpeg] 接收的是「摆正后图像」编码的 JPEG，返回的
/// 包围盒归一化到该图像的 0..1 空间，与人脸/手势覆盖层共享同一坐标基准，
/// 便于做「手持物体」的空间叠加判断。
///
/// **模型**：用官方 `yolo26n`（COCO 80 类检测）。插件 0.6.5 的原生后处理是按
/// **YOLO26 系列**设计的,与该架构最匹配。注意 ultralytics_yolo 0.6.x 已把
/// `yolo11n` 移出自动下载清单,且 v0.2.0 的旧 `yolo11n` 导出与 0.6.5 的解码器
/// 不匹配 —— 实测换 `yolo11n` 反而劣化,故仍用 `yolo26n`。
///
/// 为完全离线(陪伴机器人不依赖运行时联网),已把 `yolo26n_int8.tflite` 内置到
/// assets/models/:插件解析官方 ID `yolo26n` 时会优先从该资产拷贝到内部存储使用,
/// 找不到才回退联网下载。
///
/// **关于量化与精度**:内置的是 int8 量化模型(体积小、速度快)。int8 对相近小物体
/// (杯子/书/手机)区分略弱,若要更高精度可换**同架构的非量化版** yolo26n:
///   `yolo export model=yolo26n.pt format=tflite half=True`  (生成 fp16)
/// 导出后放到 `assets/models/yolo26n_float.tflite`,并把 [_modelCandidates] 首项
/// 改为该路径即可(int8 作为回退保留)。**切勿换成 yolo11n**:架构不匹配会更差。
///
/// **加载策略([_modelCandidates])**:按优先级依次尝试,第一个成功的即采用,
/// 任一失败(资产缺失/原生加载失败)自动回退,确保物体识别始终可用。
class ObjectEngine {
  /// 模型候选(按优先级)。默认用官方 `yolo26n`(内置 int8 资产,离线可用)。
  /// 想用非量化 yolo26n:导出 fp16 放到 assets/models/yolo26n_float.tflite,
  /// 再把该路径加到本列表最前面即可。
  static const List<String> _modelCandidates = [
    'yolo26n',
  ];

  /// 单帧最多返回的物体数（按面积降序截断）。
  final int maxObjects;

  /// 置信度阈值（低于此值的检测被过滤）。
  final double confidenceThreshold;

  /// NMS 的 IoU 阈值。
  final double iouThreshold;

  YOLO? _yolo;
  bool _initialized = false;
  String? _statusMessage;
  int _logCounter = 0;

  ObjectEngine({
    this.maxObjects = 10,
    this.confidenceThreshold = 0.35,
    this.iouThreshold = 0.45,
  });

  bool get isInitialized => _initialized;
  String? get statusMessage => _statusMessage;

  Future<void> initialize() async {
    if (_initialized) return;
    // 按优先级尝试候选模型:float 优先,失败(资产缺失/原生加载失败)逐级回退,
    // 保证物体识别始终可用。记录最后一次错误用于诊断。
    Object? lastError;
    for (final path in _modelCandidates) {
      try {
        // useGpu:false —— 陪伴机器人硬件多样,LiteRT GPU delegate 在部分设备会崩,
        // 低频检测下 CPU 足够稳妥(插件文档亦建议为稳定性关闭 GPU)。
        final yolo = YOLO(
          modelPath: path,
          task: YOLOTask.detect,
          useGpu: false,
        );
        final ok = await yolo.loadModel();
        if (!ok) {
          debugPrint('[ObjectEngine] loadModel returned false for $path');
          continue;
        }
        _yolo = yolo;
        _initialized = true;
        final quant = path.contains('int8') ? 'int8' : 'float';
        _statusMessage = '物体识别已加载 ${_shortName(path)}（$quant）';
        debugPrint('[ObjectEngine] loaded $path ($quant)');
        return;
      } catch (e) {
        lastError = e;
        debugPrint('[ObjectEngine] load $path failed: $e');
      }
    }
    _statusMessage = '物体识别模型加载失败'
        '（检查 assets/models/ 下的 yolo11n_float.tflite 是否打包）'
        '${lastError != null ? '：$lastError' : ''}';
    debugPrint('[ObjectEngine] all model candidates failed');
  }

  /// 从模型路径取一个简短可读名(去目录与扩展名),用于状态展示。
  static String _shortName(String path) =>
      path.split('/').last.replaceAll('.tflite', '');

  /// 对一张「摆正后图像」编码的 JPEG 做物体检测。
  ///
  /// [w]/[h] 为该图像的像素尺寸，用于在原生未提供归一化框时兜底归一化。
  /// 返回归一化（0..1）的物体覆盖层列表，按面积降序、至多 [maxObjects] 个。
  Future<List<ObjectOverlay>> predictJpeg(Uint8List jpg, int w, int h) async {
    final yolo = _yolo;
    if (yolo == null || !_initialized || jpg.isEmpty) return const [];

    Map<String, dynamic> res;
    try {
      res = await yolo.predict(
        jpg,
        confidenceThreshold: confidenceThreshold,
        iouThreshold: iouThreshold,
      );
    } catch (e) {
      debugPrint('[ObjectEngine] predict error: $e');
      return const [];
    }

    final dets = res['detections'];
    if (dets is! List) return const [];

    final out = <ObjectOverlay>[];
    for (final d in dets) {
      if (d is! Map) continue;
      final conf = (d['confidence'] as num?)?.toDouble() ?? 0;
      final rect = _readRect(d, w, h);
      if (rect == null || rect.width <= 0 || rect.height <= 0) continue;
      final cls = (d['className'] as String?) ?? '';
      // 跳过「人」类:人的检测/身份由专门的人脸引擎负责,YOLO 的 person 框
      // 既冗余又会被「手持」空间判定误命中,拼出「手里拿着人」这类离谱描述。
      if (cls.trim().toLowerCase() == 'person') continue;
      out.add(ObjectOverlay(
        boundingBox: rect,
        label: _displayLabel(cls),
        confidence: conf,
      ));
    }

    out.sort((a, b) => (b.boundingBox.width * b.boundingBox.height)
        .compareTo(a.boundingBox.width * a.boundingBox.height));

    if (++_logCounter <= 8 || (_logCounter % 20) == 0) {
      final dump = out
          .take(maxObjects)
          .map((o) => '${o.label}:${o.confidence.toStringAsFixed(2)}')
          .join(', ');
      debugPrint('[ObjectEngine] predict ${w}x$h dets=${dets.length} -> [$dump]');
    }

    return out.take(maxObjects).toList(growable: false);
  }

  /// 优先用原生归一化框；无效时用像素框 / (w,h) 兜底。
  Rect? _readRect(Map d, int w, int h) {
    final nb = d['normalizedBox'];
    if (nb is Map) {
      final l = (nb['left'] as num?)?.toDouble();
      final t = (nb['top'] as num?)?.toDouble();
      final r = (nb['right'] as num?)?.toDouble();
      final b = (nb['bottom'] as num?)?.toDouble();
      if (l != null && t != null && r != null && b != null && r > l && b > t) {
        return Rect.fromLTRB(
          l.clamp(0.0, 1.0),
          t.clamp(0.0, 1.0),
          r.clamp(0.0, 1.0),
          b.clamp(0.0, 1.0),
        );
      }
    }
    final bb = d['boundingBox'];
    if (bb is Map && w > 0 && h > 0) {
      final l = ((bb['left'] as num?)?.toDouble() ?? 0) / w;
      final t = ((bb['top'] as num?)?.toDouble() ?? 0) / h;
      final r = ((bb['right'] as num?)?.toDouble() ?? 0) / w;
      final b = ((bb['bottom'] as num?)?.toDouble() ?? 0) / h;
      if (r > l && b > t) {
        return Rect.fromLTRB(
          l.clamp(0.0, 1.0),
          t.clamp(0.0, 1.0),
          r.clamp(0.0, 1.0),
          b.clamp(0.0, 1.0),
        );
      }
    }
    return null;
  }

  String _displayLabel(String raw) {
    final key = raw.trim().toLowerCase();
    return _cocoZh[key] ?? raw;
  }

  /// COCO 80 类英文 → 中文。键为 YOLO 输出的英文类名（小写）。
  static const Map<String, String> _cocoZh = {
    'person': '人',
    'bicycle': '自行车',
    'car': '汽车',
    'motorcycle': '摩托车',
    'airplane': '飞机',
    'bus': '公交车',
    'train': '火车',
    'truck': '卡车',
    'boat': '船',
    'traffic light': '红绿灯',
    'fire hydrant': '消防栓',
    'stop sign': '停车标志',
    'parking meter': '停车计时器',
    'bench': '长椅',
    'bird': '鸟',
    'cat': '猫',
    'dog': '狗',
    'horse': '马',
    'sheep': '羊',
    'cow': '牛',
    'elephant': '大象',
    'bear': '熊',
    'zebra': '斑马',
    'giraffe': '长颈鹿',
    'backpack': '背包',
    'umbrella': '伞',
    'handbag': '手提包',
    'tie': '领带',
    'suitcase': '行李箱',
    'frisbee': '飞盘',
    'skis': '滑雪板',
    'snowboard': '单板滑雪',
    'sports ball': '球',
    'kite': '风筝',
    'baseball bat': '棒球棒',
    'baseball glove': '棒球手套',
    'skateboard': '滑板',
    'surfboard': '冲浪板',
    'tennis racket': '网球拍',
    'bottle': '瓶子',
    'wine glass': '酒杯',
    'cup': '杯子',
    'fork': '叉子',
    'knife': '刀',
    'spoon': '勺子',
    'bowl': '碗',
    'banana': '香蕉',
    'apple': '苹果',
    'sandwich': '三明治',
    'orange': '橙子',
    'broccoli': '西兰花',
    'carrot': '胡萝卜',
    'hot dog': '热狗',
    'pizza': '披萨',
    'donut': '甜甜圈',
    'cake': '蛋糕',
    'chair': '椅子',
    'couch': '沙发',
    'potted plant': '盆栽',
    'bed': '床',
    'dining table': '餐桌',
    'toilet': '马桶',
    'tv': '电视',
    'laptop': '笔记本电脑',
    'mouse': '鼠标',
    'remote': '遥控器',
    'keyboard': '键盘',
    'cell phone': '手机',
    'microwave': '微波炉',
    'oven': '烤箱',
    'toaster': '烤面包机',
    'sink': '水槽',
    'refrigerator': '冰箱',
    'book': '书',
    'clock': '时钟',
    'vase': '花瓶',
    'scissors': '剪刀',
    'teddy bear': '玩偶熊',
    'hair drier': '吹风机',
    'toothbrush': '牙刷',
  };

  Future<void> dispose() async {
    await _yolo?.dispose();
    _yolo = null;
    _initialized = false;
  }
}

File: xbot-native/app/src/main/java/com/xbot/android/vision/ObjectEngine.kt
package com.xbot.android.vision

import android.content.Context
import android.graphics.Bitmap
import android.graphics.RectF
import android.util.Log
import com.xbot.android.model.ObjectOverlay
import org.tensorflow.lite.Interpreter
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel
import java.io.FileInputStream
import kotlin.math.max
import kotlin.math.min

/**
 * 物体识别引擎：YOLO26（int8 TFLite）。
 *
 * 对应 Flutter 的 object_engine.dart（ultralytics_yolo 插件，YOLO26 COCO 80 类）。
 *
 * **YOLO26 端到端无 NMS 导出**：模型已内嵌 NMS，输出张量是已过滤的检测结果，
 * 形如 [1, num_detections, 6]，每行 = [x_center, y_center, width, height, conf, cls]
 *（坐标归一化到 0..1）。故本引擎**不需要再实现 NMS**，直接读结果 + 置信度过滤即可。
 *
 * 若遇到旧式 [1, 4+nc, anchors] 布局（需 NMS），[decode] 会按形状自动分支处理。
 *
 * 跳过 person 类（人的检测/身份由人脸引擎负责）。标签经 [COCO_ZH] 映射为中文。
 *
 * @param modelPath assets 中的模型路径（yolo26n_int8.tflite）
 */
class ObjectEngine(
    context: Context,
    modelPath: String,
    /** 单帧最多返回的物体数（按面积降序截断）。 */
    private val maxObjects: Int = 10,
    /** 置信度阈值（低于此值过滤）。 */
    private val confidenceThreshold: Float = 0.35f,
    /** NMS 的 IoU 阈值（仅旧式布局用到）。 */
    private val iouThreshold: Float = 0.45f,
    private val inputSize: Int = 640,
) {
    companion object {
        private const val TAG = "ObjectEngine"

        /** COCO 80 类英文（顺序即类别 index）。 */
        val COCO_LABELS = listOf(
            "person", "bicycle", "car", "motorcycle", "airplane", "bus", "train", "truck", "boat",
            "traffic light", "fire hydrant", "stop sign", "parking meter", "bench", "bird", "cat",
            "dog", "horse", "sheep", "cow", "elephant", "bear", "zebra", "giraffe", "backpack",
            "umbrella", "handbag", "tie", "suitcase", "frisbee", "skis", "snowboard", "sports ball",
            "kite", "baseball bat", "baseball glove", "skateboard", "surfboard", "tennis racket",
            "bottle", "wine glass", "cup", "fork", "knife", "spoon", "bowl", "banana", "apple",
            "sandwich", "orange", "broccoli", "carrot", "hot dog", "pizza", "donut", "cake",
            "chair", "couch", "potted plant", "bed", "dining table", "toilet", "tv", "laptop",
            "mouse", "remote", "keyboard", "cell phone", "microwave", "oven", "toaster", "sink",
            "refrigerator", "book", "clock", "vase", "scissors", "teddy bear", "hair drier", "toothbrush",
        )

        /** COCO 80 类英文 → 中文（与 Flutter object_engine.dart 一致）。 */
        val COCO_ZH: Map<String, String> = mapOf(
            "person" to "人", "bicycle" to "自行车", "car" to "汽车", "motorcycle" to "摩托车",
            "airplane" to "飞机", "bus" to "公交车", "train" to "火车", "truck" to "卡车",
            "boat" to "船", "traffic light" to "红绿灯", "fire hydrant" to "消防栓",
            "stop sign" to "停车标志", "parking meter" to "停车计时器", "bench" to "长椅",
            "bird" to "鸟", "cat" to "猫", "dog" to "狗", "horse" to "马", "sheep" to "羊",
            "cow" to "牛", "elephant" to "大象", "bear" to "熊", "zebra" to "斑马",
            "giraffe" to "长颈鹿", "backpack" to "背包", "umbrella" to "伞", "handbag" to "手提包",
            "tie" to "领带", "suitcase" to "行李箱", "frisbee" to "飞盘", "skis" to "滑雪板",
            "snowboard" to "单板滑雪", "sports ball" to "球", "kite" to "风筝",
            "baseball bat" to "棒球棒", "baseball glove" to "棒球手套", "skateboard" to "滑板",
            "surfboard" to "冲浪板", "tennis racket" to "网球拍", "bottle" to "瓶子",
            "wine glass" to "酒杯", "cup" to "杯子", "fork" to "叉子", "knife" to "刀",
            "spoon" to "勺子", "bowl" to "碗", "banana" to "香蕉", "apple" to "苹果",
            "sandwich" to "三明治", "orange" to "橙子", "broccoli" to "西兰花", "carrot" to "胡萝卜",
            "hot dog" to "热狗", "pizza" to "披萨", "donut" to "甜甜圈", "cake" to "蛋糕",
            "chair" to "椅子", "couch" to "沙发", "potted plant" to "盆栽", "bed" to "床",
            "dining table" to "餐桌", "toilet" to "马桶", "tv" to "电视", "laptop" to "笔记本电脑",
            "mouse" to "鼠标", "remote" to "遥控器", "keyboard" to "键盘", "cell phone" to "手机",
            "microwave" to "微波炉", "oven" to "烤箱", "toaster" to "烤面包机", "sink" to "水槽",
            "refrigerator" to "冰箱", "book" to "书", "clock" to "时钟", "vase" to "花瓶",
            "scissors" to "剪刀", "teddy bear" to "玩偶熊", "hair drier" to "吹风机", "toothbrush" to "牙刷",
        )
    }

    private var interpreter: Interpreter? = null
    var isInitialized = false
        private set

    init {
        try {
            val modelBuffer = loadModelFile(context, modelPath)
            val itp = Interpreter(modelBuffer, Interpreter.Options().setNumThreads(4))
            interpreter = itp
            isInitialized = true
            val inShape = itp.getInputTensor(0).shape().toList()
            val outShape = itp.getOutputTensor(0).shape().toList()
            Log.i(TAG, "YOLO26 已加载，inShape=$inShape outShape=$outShape")
        } catch (e: Exception) {
            Log.e(TAG, "物体识别模型加载失败: ${e.message}")
            isInitialized = false
        }
    }

    /**
     * 对一张摆正后的 bitmap 做物体检测。返回归一化（0..1）物体列表，按面积降序、
     * 至多 [maxObjects] 个。
     */
    fun detect(bitmap: Bitmap): List<ObjectOverlay> {
        val itp = interpreter ?: return emptyList()
        return try {
            // YOLO 输入：640x640 RGB，归一化 /255（letterbox 缩放保持比例）。
            val letterboxed = letterbox(bitmap, inputSize)
            val input = buildInput(letterboxed)
            // 输出布局自适应：先按 outShape(0) 判断。
            val outShape = itp.getOutputTensor(0).shape()
            val detections = when {
                // 端到端无 NMS：[1, N, 6] 或 [1, N, 4+1+1]
                outShape.size == 3 && outShape[2] <= 6 && outShape[1] > outShape[2] -> {
                    val out = Array(1) { Array(outShape[1]) { FloatArray(outShape[2]) } }
                    itp.run(arrayOf(input), out)
                    decodeEndToEnd(out[0], letterboxed.scaleX, letterboxed.scaleY, letterboxed.padX, letterboxed.padY)
                }
                // 旧式：[1, 4+nc, anchors] 需 NMS
                outShape.size == 3 && outShape[1] >= 4 + 80 -> {
                    val nc = outShape[1] - 4
                    val anchors = outShape[2]
                    val out = Array(1) { Array(outShape[1]) { FloatArray(anchors) } }
                    itp.run(arrayOf(input), out)
                    decodeClassic(out[0], nc, anchors, letterboxed.scaleX, letterboxed.scaleY, letterboxed.padX, letterboxed.padY)
                }
                else -> {
                    Log.w(TAG, "未知输出布局 outShape=$outShape，跳过")
                    emptyList()
                }
            }
            detections
                .filterNot { it.label.equals("person", ignoreCase = true) || it.label == "人" }
                .sortedByDescending { it.boundingBox.width() * it.boundingBox.height() }
                .take(maxObjects)
        } catch (e: Exception) {
            Log.e(TAG, "detect 异常: ${e.message}")
            emptyList()
        }
    }

    /** 端到端无 NMS 解码：每行 [cx, cy, w, h, conf, cls]，坐标相对 letterbox 后的 640x640。 */
    private fun decodeEndToEnd(
        rows: Array<FloatArray>,
        scaleX: Float,
        scaleY: Float,
        padX: Float,
        padY: Float,
    ): List<ObjectOverlay> {
        val out = ArrayList<ObjectOverlay>()
        for (row in rows) {
            if (row.size < 6) continue
            val conf = row[4]
            if (conf < confidenceThreshold) continue
            val clsIdx = row[5].toInt()
            // 反 letterbox：去掉 padding + 缩放回原图归一化空间。
            val cx = (row[0] - padX) * scaleX
            val cy = (row[1] - padY) * scaleY
            val w = row[2] * scaleX
            val h = row[3] * scaleY
            val left = (cx - w / 2f).coerceIn(0f, 1f)
            val top = (cy - h / 2f).coerceIn(0f, 1f)
            val right = (cx + w / 2f).coerceIn(0f, 1f)
            val bottom = (cy + h / 2f).coerceIn(0f, 1f)
            val rawLabel = COCO_LABELS.getOrElse(clsIdx) { "object_$clsIdx" }
            out.add(
                ObjectOverlay(
                    boundingBox = RectF(left, top, right, bottom),
                    label = COCO_ZH[rawLabel] ?: rawLabel,
                    confidence = conf,
                )
            )
        }
        return out
    }

    /** 旧式 [4+nc, anchors] 解码 + NMS。 */
    private fun decodeClassic(
        data: Array<FloatArray>,
        nc: Int,
        anchors: Int,
        scaleX: Float,
        scaleY: Float,
        padX: Float,
        padY: Float,
    ): List<ObjectOverlay> {
        // data[row][anchor]，row = 0..3 是 xywh，row = 4..4+nc-1 是各类置信度。
        val candidates = ArrayList<ObjectOverlay>()
        for (a in 0 until anchors) {
            // 找最高类别置信度。
            var bestCls = -1
            var bestConf = confidenceThreshold
            for (c in 0 until nc) {
                val conf = data[4 + c][a]
                if (conf > bestConf) {
                    bestConf = conf
                    bestCls = c
                }
            }
            if (bestCls < 0) continue
            val cx = (data[0][a] - padX) * scaleX
            val cy = (data[1][a] - padY) * scaleY
            val w = data[2][a] * scaleX
            val h = data[3][a] * scaleY
            val left = (cx - w / 2f).coerceIn(0f, 1f)
            val top = (cy - h / 2f).coerceIn(0f, 1f)
            val right = (cx + w / 2f).coerceIn(0f, 1f)
            val bottom = (cy + h / 2f).coerceIn(0f, 1f)
            val rawLabel = COCO_LABELS.getOrElse(bestCls) { "object_$bestCls" }
            candidates.add(
                ObjectOverlay(
                    boundingBox = RectF(left, top, right, bottom),
                    label = COCO_ZH[rawLabel] ?: rawLabel,
                    confidence = bestConf,
                )
            )
        }
        return nms(candidates, iouThreshold)
    }

    /** NMS（非极大值抑制）。 */
    private fun nms(boxes: List<ObjectOverlay>, iouThreshold: Float): List<ObjectOverlay> {
        val sorted = boxes.sortedByDescending { it.confidence }.toMutableList()
        val result = ArrayList<ObjectOverlay>()
        while (sorted.isNotEmpty()) {
            val best = sorted.removeAt(0)
            result.add(best)
            sorted.removeAll { iou(best.boundingBox, it.boundingBox) > iouThreshold }
        }
        return result
    }

    private fun iou(a: RectF, b: RectF): Float {
        val left = max(a.left, b.left)
        val top = max(a.top, b.top)
        val right = min(a.right, b.right)
        val bottom = min(a.bottom, b.bottom)
        val interW = right - left
        val interH = bottom - top
        if (interW <= 0 || interH <= 0) return 0f
        val inter = interW * interH
        val union = a.width() * a.height() + b.width() * b.height() - inter
        return if (union <= 0f) 0f else inter / union
    }

    /** letterbox 信息：缩放后的 bitmap + 反变换系数。 */
    private data class Letterbox(
        val bitmap: Bitmap,
        val scaleX: Float, // 640空间坐标 → 原图归一化：scaleX = 1 / (有效宽 / 原宽)
        val scaleY: Float,
        val padX: Float,   // 640空间下的左 padding（归一化）
        val padY: Float,
    )

    /** 等比缩放到 640x640，短边填充灰色(114)，保持比例。记录反变换参数。 */
    private fun letterbox(src: Bitmap, size: Int): Letterbox {
        val scale = min(size.toFloat() / src.width, size.toFloat() / src.height)
        val newW = (src.width * scale).toInt()
        val newH = (src.height * scale).toInt()
        val scaled = Bitmap.createScaledBitmap(src, newW, newH, true)
        val out = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
        // 填充灰色 114。
        val canvas = android.graphics.Canvas(out)
        canvas.drawColor(android.graphics.Color.rgb(114, 114, 114))
        val padX = (size - newW) / 2f
        val padY = (size - newH) / 2f
        canvas.drawBitmap(scaled, padX, padY, null)
        // 反变换：640空间坐标(x) → 原图归一化：(x - padX)/(newW) ... 但乘以 1/scale 得原图像素，再 /原宽。
        // 归一化空间下：scaleX = (原图宽) / (newW) （把 640 有效宽还原成原图宽占比）。
        val scaleX = src.width.toFloat() / newW
        val scaleY = src.height.toFloat() / newH
        return Letterbox(out, scaleX, scaleY, padX / size, padY / size)
    }

    /** 构造 YOLO 输入：1x640x640x3 float32，归一化 /255。 */
    private fun buildInput(lb: Letterbox): Array<Array<Array<FloatArray>>> {
        val pixels = IntArray(size = inputSize * inputSize)
        lb.bitmap.getPixels(pixels, 0, inputSize, 0, 0, inputSize, inputSize)
        val out = Array(inputSize) {
            Array(inputSize) { FloatArray(3) }
        }
        for (y in 0 until inputSize) {
            for (x in 0 until inputSize) {
                val px = pixels[y * inputSize + x]
                out[y][x][0] = ((px shr 16) and 0xFF) / 255f
                out[y][x][1] = ((px shr 8) and 0xFF) / 255f
                out[y][x][2] = (px and 0xFF) / 255f
            }
        }
        return arrayOf(out)
    }

    fun close() {
        interpreter?.close()
        interpreter = null
        isInitialized = false
    }

    private fun loadModelFile(context: Context, modelPath: String): MappedByteBuffer {
        val fd = context.assets.openFd(modelPath)
        FileInputStream(fd.fileDescriptor).use { fis ->
            return fis.channel.map(
                FileChannel.MapMode.READ_ONLY,
                fd.startOffset,
                fd.declaredLength,
            ).order(ByteOrder.nativeOrder()) as MappedByteBuffer
        }
    }
}
